util/svdgen.py

```diff
@@ -0,0 +1,604 @@
+    layouts, register definitions, and similar microcontroller fields.
+
+    The input HJSON must be a top-level configuration (earlgrey) and the
+    full set of IP modeule definitions for that configuration. These two
+    will be parsed and converted into a single XML tree with a top-level
+    "<device>" node. Both the top configuration and IP modules must have
+    been previously validated. (see ??? and `reggen.validate()`)
+
+    If `validate` is true then the resulting SVD file is validated using
+    the `pysvd` module.
+
+    ¹ http://www.keil.com/pack/doc/CMSIS/SVD/html/index.html
+    ² https://github.com/rust-embedded/svd2rust"""
+
+    root = device(top, ips, version)
+
+    # Simply constructing a Device is enough to run the pysvd parser and
+    # validate the structure of the XML tree.
+    if validate:
+        pysvd.element.Device(root)
+
+    # Workaround for limitations of SVD/pysvd. See comment in `cpu()`.
+    root.find('.cpu/name').text = 'RISCV'
+
+    beautify(root)
+    return root
+
+def write_to_file(device: ET.Element, copyright: [str], output):
+    """Write out the generated SVD <device> element and its children to a
+    file in XML format. The SVD contents are preceeded by a comment with
+    the given copyright notice and a warning that the file was generated
+    by this script."""
+
+    def write_element(element, xml):
+        et = ET.ElementTree(element)
+        et.write(output, encoding='unicode', xml_declaration=xml)
+
+    comment = ET.Comment('\n'.join(('', *copyright, GENERATED_NOTICE, '')))
+    comment.tail = '\n'
+
+    # The python3 ElementTree API doesn't natively support a top-level
+    # documentation comment. Fake it for reader clarity.
+    write_element(comment, True)
+    write_element(device, False)
+
+def validate(top: hjson, ips: {"name": hjson}) -> bool:
+    """Validate the combined top-level HJSON and IP module definitions.
+    This checks for required fields in all the HJSON and ensures that a
+    corresponding IP block exists for all modules in `top`.
+
+    This assumes each individual IP module was previously validated using
+    the `reggen.validate.validate()` function."""
+
+    for field in ['name', 'datawidth', 'module']:
+        if not field in top:
+            yield 'missing required field %s in top hjson' % field
+
+    for (i, module) in enumerate(top['module']):
+        name = module.get('name') or ordinal(i)
+
+        for field in ['name', 'type', 'base_addr']:
+            if not field in module:
+                yield 'missing field "%s" in %s module' % (field, name)
+
+        tipe = module.get('type')
+        if tipe != None and not tipe in ips:
+            yield 'module "%s" references missing IP hjson: "%s"' % (name, tipe)
+
+    for ip in ips.values():
+        if ip['regwidth'] != top['datawidth']:
+            yield 'register width of module "%s" does not top hjson' % ip['name']
+
+
+def test():
```

Comment:
What about committing some test hjson and the corresponding generated svd and do structural comparison?

Reply:
With the added tests, I split `svdgen` up into a layout similar to `topgen`'s, with subdirectory containing the test and library code.

I added a [series of tests](https://github.com/nholstein/opentitan/blob/af065293154182ce654645bcdd98e01f51b8f265/util/svdgen/test.py#L88-L429) which generate an SVD <peripheral> from a small HJSON definition and compares the resulting XML tree against the expected results. The test data is a combination of registers pulled from hw/ip, and some fabricated.